### PR TITLE
fix the index of header when build rest client

### PR DESCRIPTION
**Pull request by neomerckx**
the client header is not set properly because of missing the i++.

### CODE REVIEW

**Review by jasontedor — CHANGES_REQUESTED** [review]
This PR needs a test that fails without this fix.

**Comment by slixurd** [comment]
@jasontedor 
Already add the unit test for this fix.
There is no check that the numbers of the headers to set  and the numbers of the headers in rest client are equal, because there is no public method to get the numbers of the headers in rest client .
But The Check in RestClient.builder() will result in NullPointerException when the headers are not set properly.

**Comment by rjernst** [comment]
@jasontedor Are you happy with the test which was added?

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteBuildRestClientTests.java [hunk 1]
```diff
@@ -49,4 +49,29 @@ public void testBuildRestClient() throws Exception {
...
     }
+    
+    public void testBuildRestClientWithHeader() throws Exception {
+        Map<String, String> headers = new HashMap<>();
+        int headerNumbers = randomIntBetween(2, 5);
+        for (int i = 0; i < headerNumbers; i++) {
+            headers.put("header" + i, Integer.toString(i));
+        }
+        assertEquals(headerNumbers, headers.size());
+        System.out.println(headers.toString());
```
Comment (jasontedor) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteBuildRestClientTests.java]: Can you remove this debugging output?

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteBuildRestClientTests.java [hunk 2]
```diff
@@ -49,4 +49,29 @@ public void testBuildRestClient() throws Exception {
...
+        RemoteInfo remoteInfo = new RemoteInfo("https", "localhost", 9200, new BytesArray("ignored"), null, null, headers,
+            RemoteInfo.DEFAULT_SOCKET_TIMEOUT, RemoteInfo.DEFAULT_CONNECT_TIMEOUT);
+        long taskId = randomLong();
+        List<Thread> threads = synchronizedList(new ArrayList<>());
+        RestClient client = TransportReindexAction.buildRestClient(remoteInfo, taskId, threads);
```
Comment (jasontedor) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteBuildRestClientTests.java]: This test should assert that the headers are correct.

**Review by jasontedor — CHANGES_REQUESTED** [review]
Sorry for the slow reply, I left some comments regarding the test.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportReindexAction.java [hunk 3]
```diff
@@ -204,7 +204,7 @@ static RestClient buildRestClient(RemoteInfo remoteInfo, long taskId, List<Threa
         Header[] clientHeaders = new Header[remoteInfo.getHeaders().size()];
         int i = 0;
         for (Map.Entry<String, String> header : remoteInfo.getHeaders().entrySet()) {
-            clientHeaders[i] = new BasicHeader(header.getKey(), header.getValue());
+            clientHeaders[i++] = new BasicHeader(header.getKey(), header.getValue());
```
Comment (nik9000) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportReindexAction.java]: Yeah, that looks like the right fix....

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteBuildRestClientTests.java (hunk 2 above)
Comment (nik9000) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteBuildRestClientTests.java]: I agree with @jasontedor. I'd be fine with a adding `RestClient#getDefaultHeaders` which returns a copy of the `defaultHeaders` array so we can assert that they are correct. @javanna, what do you think of that?

Also, I don't think there is any need to assert on the threads in this test. The test above is careful about that already.

**Review by nik9000 — CHANGES_REQUESTED** [review]

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteBuildRestClientTests.java (hunk 2 above)
Comment (javanna) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteBuildRestClientTests.java]: I am ok with what you propose Nik!

**Comment by dakrone** [comment]
@slixurd is this something you're still interested in pursuing? Can you address the feedback?

**Comment by rjernst** [comment]
@slixurd As this PR seems to be stalled, I am going to close it. I have opened #26937 with the suggested tests, so that we can get this fix into 6.0. I hope you understand.

**Pull request closed without merging**